### PR TITLE
Better @DataBoundConstructor hygiene for ParameterDefinition’s

### PR DESCRIPTION
**Pull request by jglick**
We generally advocate that a `@DataBoundConstructor` include only genuinely mandatory fields, and that anything with a sensible default value be put in a `@DataBoundSetter`. Not only did core-defined parameter types not comply, the API signature of `ParameterDefinition` made it actively difficult for subclasses to comply, because the `description` was `final` and could only be passed in the `super` constructor.

Inspired by https://github.com/jenkinsci/file-parameters-plugin/pull/10. In that case I found a workaround (overriding `getDescription`), allowing a more compact form of the parameter definition which is particularly desirable in Declarative Pipeline: if only the `name` is truly mandatory, you would prefer to be able to write

```groovy
parameters {
  theParamType 'myParamName'
}
```

rather than

```groovy
parameters {
  theParamType(name: 'myParamName')
}
```

Also nice in functional tests:

```diff
-project.addProperty(new ParametersDefinitionProperty(new StringParameterDefinition("xxx", null)));
+project.addProperty(new ParametersDefinitionProperty(new StringParameterDefinition("xxx")));
```

### Proposed changelog entries

* Relaxed `ParameterDefinition` and subclass constructor signatures to allow optional attributes like `description` to be defined in setters.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/model/ChoiceParameterDefinition.java
Comment (jglick) [inline core/src/main/java/hudson/model/ChoiceParameterDefinition.java:60]: Looked complicated to attempt right now: this constructor does not bind the default value for some reason!

File: core/src/main/java/hudson/model/ParameterDefinition.java
Comment (jglick) [inline core/src/main/java/hudson/model/ParameterDefinition.java:97]: Ancient, probably obsolete comments.

File: core/src/main/java/hudson/model/RunParameterDefinition.java
Comment (jglick) [inline core/src/main/java/hudson/model/RunParameterDefinition.java:66]: A bit more complex in this case. Also less helpful since you would still have two mandatory attributes.

File: core/src/main/java/hudson/model/PasswordParameterDefinition.java
Comment (jglick) [inline core/src/main/java/hudson/model/PasswordParameterDefinition.java:57]: Looked tricky due to the historical confusion over the attribute name: `defaultValue` vs. `defaultValueAsSecret`.

File: core/src/main/java/hudson/model/BooleanParameterDefinition.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/BooleanParameterDefinition.java:44]:
```suggestion
     * @since TODO
```

Plus elsewhere, for compatibility with https://github.com/jenkinsci/jenkins/blob/71674c92c9118ff7cd4d34f508e7da9b2571399d/update-since-todo.sh#L13

Comment (jglick) [reply core/src/main/java/hudson/model/BooleanParameterDefinition.java:44]: OK, I could not remember which it was!

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by jglick** [comment]
Made material changes, leaving it up for another round of review.

**Comment by timja** [comment]
> Made material changes, leaving it up for another round of review.

`hudson.model.ChoiceParameterDefinitionTest.checkValue_Null` and `hudson.model.ChoiceParameterDefinitionTest.createNullChoice` are now failing
https://github.com/jenkinsci/jenkins/pull/5275/checks?check_run_id=1907987287

**Review by timja — APPROVED** [review]

File: core/src/main/java/hudson/model/FileParameterDefinition.java
Comment (bitwiseman) [inline core/src/main/java/hudson/model/FileParameterDefinition.java:57]: Should we mark the other constructors as deprecated? (Repeat this for other parameter classes.)
```suggestion
    @Deprecated
    public FileParameterDefinition(String name, String description) {
```

**Review by bitwiseman — APPROVED** [review]

File: core/src/main/java/hudson/model/FileParameterDefinition.java
Comment (jglick) [reply core/src/main/java/hudson/model/FileParameterDefinition.java:57]: No strong opinion, but I did not see any particular reason to deprecate these constructors if they are working fine for functional tests and the like. I deprecated the _super_ constructor in `ParameterDefinition` as a reminder that subtypes (in plugin `src/main/java/`) ought to offer the simplified constructors.

File: core/src/main/java/hudson/model/StringParameterValue.java
Comment (timja) [reply core/src/main/java/hudson/model/StringParameterValue.java:52]: there's a few failures, https://github.com/jenkinsci/bom/pull/511 e.g. https://github.com/jenkinsci/bom/pull/511/checks?check_run_id=2535377510

Comment (jtnord) [reply core/src/main/java/hudson/model/StringParameterValue.java:52]: was just about to comment the same :)

SO should we fix all the tests, or revert this specific part?

@jglick do you recall the reason for this specific change?

Comment (jtnord) [reply core/src/main/java/hudson/model/StringParameterValue.java:52]: > Treat empty ParameterDefinition.description & StringParameterDefinition.defaultValue like null, for better /directive-generator/ behavior

Which seems to be the inverse of this change?  Ie `""` should go to `null` but this makes `null` -> `""`

ie this should be `Util.fixEmpty` ?

**Pull request merged**
